edk2basetools/VfrCompiler/IfrTree.py

```diff
@@ -0,0 +1,1528 @@
+from antlr4 import *
+from edk2basetools.VfrCompiler.IfrCtypes import *
+from edk2basetools.VfrCompiler.IfrPreProcess import *
+from edk2basetools.VfrCompiler.IfrFormPkg import *
+
+# Ifr related Info -> ctypes obj
+# 　conditional Info
+# Structure Info
+
+VFR_COMPILER_VERSION = "2.01 (UEFI 2.4)"
+BUILD_VERSION = "Developer Build based on Revision: Unknown"
+
+
+class IfrTreeNode:
+    def __init__(self, OpCode=None, Data=None, Buffer=None, Position=None) -> None:
+        self.OpCode = OpCode
+        self.Data = Data
+        self.Buffer = Buffer
+        self.Position = Position
+        self.Condition = None
+        self.Expression = None
+        self.Dict = {}
+        self.Offset = None
+        self.Parent = None
+        self.Child = []
+        self.Level = -1
+
+    def hasCondition(self) -> bool:
+        if self.Condition == None:
+            return False
+        else:
+            return True
+
+    def hasChild(self) -> bool:
+        if self.Child == []:
+            return False
+        else:
+            return True
+
+    def isFinalChild(self) -> bool:
+        ParTree = self.Parent
+        if ParTree:
+            if ParTree.Child[-1] == self:
+                return True
+        return False
+
+    def insertChild(self, NewNode, pos: int = None) -> None:
+        if NewNode != None:
+            if not pos:
+                self.Child.append(NewNode)
+            else:
+                self.Child.insert(pos, NewNode)
+
+            NewNode.Parent = self
+
+    # lastNode.insertRel(newNode)
+    def insertRel(self, newNode) -> None:
+        if self.Parent:
+            parentTree = self.Parent
+            new_index = parentTree.Child.index(self) + 1
+            parentTree.Child.insert(new_index, newNode)
+        self.NextRel = newNode
+        newNode.LastRel = self
+
+    def deleteNode(self, deletekey: str) -> None:
+        FindStatus, DeleteTree = self.FindNode(deletekey)
+        if FindStatus:
+            parentTree = DeleteTree.Parent
+            lastTree = DeleteTree.LastRel
+            nextTree = DeleteTree.NextRel
+            if parentTree:
+                index = parentTree.Child.index(DeleteTree)
+                del parentTree.Child[index]
+            if lastTree and nextTree:
+                lastTree.NextRel = nextTree
+                nextTree.LastRel = lastTree
+            elif lastTree:
+                lastTree.NextRel = None
+            elif nextTree:
+                nextTree.LastRel = None
+            return DeleteTree
+        else:
+            print("Could not find the target tree")
```

Comment:
print, don't log. Also this else statement is unecessary. Just shift the print (to be log) statement to the left by a tab.